### ISSUE TITLE
in BertForMaskedLM, why masked_tokens_mask = labels > 0?

### ISSUE THREAD

**Issue by JadenNeal** [mosaicml/examples#457]
Hi, thank you for your outstanding work. 

Recently, I am using your [bert model](https://github.com/mosaicml/examples/blob/main/examples/benchmarks/bert/src/bert_layers.py#L854) for pre-training. But in line 854, i found that

```python
https://github.com/mosaicml/examples/blob/main/examples/benchmarks/bert/src/bert_layers.py
if labels is None:
    masked_tokens_mask = None
else:
    masked_tokens_mask = labels > 0
```

For a tokenizer, the word number is [0,vocab-1]. For example, Among them, 0-4 correspond to [UNK], [PAD], [CLS], [SEP], [MASK]. 

Therefore, for example, for an input text, the result of tokenization is [2, 5, 6, 6, 7, 8, 3, 1, 1] (note that 1,2,3 correspond to [PAD], [CLS], [SEP]). In the masking process(ref to huggingface [DataCollatorForLanguageModeling](https://github.com/huggingface/transformers/blob/main/src/transformers/data/data_collator.py#L831)), we have

(1) Each token has a probability of p being selected.
```python
probability_matrix = torch.full(labels.shape, self.mlm_probability)
```
(2) In the meantime, special_tokens generates a mask of [1,0,0,0,0,0,1,1,1]
```python
if special_tokens_mask is None:
    special_tokens_mask = [
        self.tokenizer.get_special_tokens_mask(val, already_has_special_tokens=True) for val in labels.tolist()
    ]
    special_tokens_mask = torch.tensor(special_tokens_mask, dtype=torch.bool)
else:
    special_tokens_mask = special_tokens_mask.bool()
```
(3) After the first two steps, only normal tokens are retained. For example, the result is [0,0,1,0,1,1,0,0,0], where the position of 1 indicates the position of the subsequent operation, but the special token will definitely not be operated. 
**so we can get masked_indices**
(4) so we can mask labels with -100. 
```python
labels[~masked_indices] = -100  # We only compute loss on masked tokens
```
Therefore, labels become [-100, -100, 6, -100, 7, 8, -100, -100,-100]. 
(4-1) Then, there is an 80% probability that the position 1 will be replaced by [MASK], a 10% probability that it will be replaced by a random word (any one of [0, vocab-1]), and a 10% probability that it will remain unchanged. 
(4-2) Assume that the final labels become [-100, -100, 4, -100, 0, 8, -100, -100,-100]. It can be seen that the original 6 is replaced by [MASK]=4. The original 7 is replaced by a random word ([UNK]=0). And 8 remains unchanged.

Therefore, the so-called masked_tokens_mask is [0,0,1,0,1,1,0,0,0] (=masked_indices), which corresponds one-to-one to labels[-100, -100, 4, -100, 0, 8, -100, -100,-100]. 
Therefore, in order to get [0,0,1,0,1,1,0,0,0], **it should be labels>=0 instead of labels>0.**

In [your comments](https://github.com/mosaicml/examples/blob/main/examples/benchmarks/bert/src/bert_layers.py#L841), you said that
```python
# Indices should be in `[-100, 0, ..., config.vocab_size]` (see
# `input_ids` docstring) Tokens with indices set to `-100` are ignored
# (masked), the loss is only computed for the tokens with labels in `[0,
# ..., config.vocab_size]`
```
Note that `the loss is only computed for the tokens with labels in [0, ..., config.vocab_size]`, but you set `labels > 0` to remove 0, so you did not compute loss with label 0.

It is weird and contradictory.

**My question is, Why do you use a stricter mask strategy for labels here and why do you remove 0?**

Thank you.

**Comment by dakinggg** [comment]
I believe 0 is pad token
```
In [1]: import transformers
bertt
In [2]: bertt = transformers.AutoTokenizer.from_pretrained('google-bert/bert-base-uncased')

In [3]: bertt.decode(0)
Out[3]: '[PAD]'
```

**Comment by JadenNeal** [comment]
> I believe 0 is pad token
> 
> ```
> In [1]: import transformers
> bertt
> In [2]: bertt = transformers.AutoTokenizer.from_pretrained('google-bert/bert-base-uncased')
> 
> In [3]: bertt.decode(0)
> Out[3]: '[PAD]'
> ```

Thank you for your reply. 

but I am still confused. Special tokens have been processed in [DataCollatorForLanguageModeling](https://github.com/huggingface/transformers/blob/main/src/transformers/data/data_collator.py#L831), which means **all special tokens have been replaced with -100**. 

All values ​​other than -100 in labels are selected as masks.

Therefore, even if 0 appears in labels, it is the result of a 10% probability of being replaced by a random word. For the model, it is still necessary to calculate the loss of the position labels=0, because this position may be another token, but it is just randomly replaced.

Moreover, using the constraint of labels>0 requires a specific tokenizer, that is, [PAD]=0. What if [PAD]=1? In this case, in order to achieve your goal, do you need to change it to labels>=1?

**Comment by dakinggg** [comment]
to be honest i haven't looked at this code in a long time. that being said, I think I see what you mean. You're probably correct that the modeling code assumes a specific tokenizer and may not be robust to all other tokenizers.

But I believe the 10% replacement occurs only on the _input_ tokens (https://github.com/huggingface/transformers/blob/f5620a76344595dbc7c9cff97bbd1edc1696854d/src/transformers/data/data_collator.py#L859), as the labels are a clone of the inputs before any of the 80/10/10 bit. So I think in the case where pad token is 0, 0 should never occur in the labels anyway.

**Comment by JadenNeal** [comment]
Thank you. 
Sorry, I overlooked the relationship between input and labels.
Yes, labels will only fill the non-selected positions with -100, and will not use 80/10/10 bit.
In fact, your operation is correct. As long as the special mask is pre-processed, [PAD] will not appear in labels. Generally speaking, in bert's tokenizer, special tokens are in the range of 0-4, so labels>0 and labels>4 have the same effect.
